agent/tracer.go

```diff
@@ -1,31 +1,117 @@
 package agent
 
 import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
 	"net/url"
+	"sync"
 	"time"
 
+	"github.com/elastic/hey-apm/conv"
+	"github.com/elastic/hey-apm/strcoll"
+
 	"go.elastic.co/apm"
 	apmtransport "go.elastic.co/apm/transport"
 )
 
-func Tracer(logger apm.Logger, serverUrl, serverSecret string, maxSpans int) *apm.Tracer {
-	tracer := apm.DefaultTracer
-	tracer.SetLogger(logger)
-	tracer.SetMetricsInterval(0) // disable metrics
-	transport := tracer.Transport.(*apmtransport.HTTPTransport)
+type Tracer struct {
+	*apm.Tracer
+	TransportStats *transportStats
+}
+
+type transportStats struct {
+	Accepted  float64
+	TopErrors []string
+}
+
+func (t Tracer) Close() {
+	t.Tracer.Close()
+	rt := t.Transport.(*apmtransport.HTTPTransport).Client.Transport.(*roundTripper)
+	rt.wg.Wait()
+	close(rt.c)
+}
+
+func NewTracer(logger apm.Logger, serverUrl, serverSecret string, maxSpans int) *Tracer {
+
+	goTracer := apm.DefaultTracer
+	goTracer.SetLogger(logger)
+	goTracer.SetMetricsInterval(0) // disable metrics
+	goTracer.SetSpanFramesMinDuration(1 * time.Nanosecond)
+	goTracer.SetMaxSpans(maxSpans)
+
+	transport := goTracer.Transport.(*apmtransport.HTTPTransport)
 	transport.SetUserAgent("hey-apm")
 	if serverSecret != "" {
 		transport.SetSecretToken(serverSecret)
 	}
 	if serverUrl != "" {
 		u, err := url.Parse(serverUrl)
 		if err != nil {
-			logger.Errorf("invalid apm server url:", err)
+			panic(err)
 		}
 		transport.SetServerURL(u)
 	}
+	rt := &roundTripper{c: make(chan []byte, 0)}
+	transport.Client.Transport = rt
 
-	tracer.SetSpanFramesMinDuration(1 * time.Nanosecond)
-	tracer.SetMaxSpans(maxSpans)
+	tracer := &Tracer{goTracer, &transportStats{}}
+
+	go func() {
+		for {
+			select {
+			case response := <-rt.c:
```

Comment:
These 3 lines could be for `for response := range rt.c {` ?

Reply:
yeah, leftover